Implement RIDE_ENTRY_FLAG_RIDER_CONTROLS_SPEED

X123M3-256 opened the pull request:
This PR adds a new vehicle flag for vehicles with rider operated brakes. See #16825 for detailed description.

duncanspumpkin commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -2443,6 +2443,16 @@ void Vehicle::UpdateWaitingToDepart()
+    if (GetRideEntry()->flags & RIDE_ENTRY_FLAG_RIDER_CONTROLS_SPEED && num_peeps != 0)
+    {
+        // There are 64 possible settings for the rider speed preferences - pick one at random
+        rider_speed_preference = (scenario_rand() & 0x3F);
```
I would instead tie it to the sprite_index of the rider. No need to make it random and then riders will be consistently be slow/fast

duncanspumpkin commented on src/openrct2/ride/Vehicle.h:
```diff
@@ -143,6 +143,7 @@ struct Vehicle : EntityBase
             int8_t ferris_wheel_var_0;
             int8_t ferris_wheel_var_1;
         };
+        int16_t rider_speed_preference;
```
Do not add to the union make this a new variable. All variables are to be upper CamelCase

X123M3-256 replied:
Won't this cause issues with saving? Would I then have to change the save format to storethe new value?

duncanspumpkin replied:
Having to add an extra field to serialisation isn't much of an issue but yeah you could get away without the property if you tie it to sprite index

duncanspumpkin commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -8028,6 +8038,109 @@ bool Vehicle::UpdateTrackMotionForwardsGetNewTrack(uint16_t trackType, Ride* cur
+{
+    int32_t mode = rider_speed_preference & 0x3F;
+
+    int32_t straightTargetSpeed[] = {
```
Make this into a struct for the settings. Move the struct and settings to a dedicated file. I would also consider moving this function to that file. Vehicle.cpp is already far too busy as is and since this is all just for one item it will make for an easy to define interface.

duncanspumpkin commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -8028,6 +8038,109 @@ bool Vehicle::UpdateTrackMotionForwardsGetNewTrack(uint16_t trackType, Ride* cur
+        654840, 663516, 671861, 679918, 687726, 695316, 702715,  709947,  717033,  723991,  730839,  737590,  744260,
+        750861, 757404, 763901, 770363, 776801, 783223, 789640,  796062,  802500,  808962,  815459,  822002,  828603,
+        835273, 842024, 848872, 855830, 862916, 870148, 877547,  885137,  892945,  901002,  909347,  918023,  927086,
+        936600, 946650, 957343, 968817, 981264, 994950, 1010270, 1027844, 1048741, 1075064, 1112052, 1182524,
```
I don't like having the speeds defined like this but i know why you have done it. It would be useful to have the speeds in terms of mph next to it. Some of these speeds are a bit odd 5.956mph, 18.044mph. wouldn't it be neater to just go up in some simple mph increment.

X123M3-256 replied:
> I don't like having the speeds defined like this but i know why you have done it. It would be useful to have the speeds in terms of mph next to it.

I couldn't work out how to convert between these integer values and the actual speed.

>  Some of these speeds are a bit odd 5.956mph, 18.044mph. wouldn't it be neater to just go up in some simple mph increment.

The speeds are sampled from the inverse CDF of the normal distribution. This is done so that if speeds are selected from the table with equal probability, the distribution of rider speeds will be normal. This is done to make extreme values uncommon, because really slow riders hold everyone else up (but I don't want to have no slow values at all).

I'm still not sure if a normal distribution is the right choice, as it might be better for the distribution to be skewed towards the upper end of the range (many fast riders will get held up by the rider in front, so if you want the player to observe a range of speeds, you might need to have disproportionately more fast riders than slow ones).

duncanspumpkin replied:
The speeds will be afair (value / 65536)_mph. (I know this is definitely correct for openloco pretty sure openrct2 is the same)

X123M3-256 replied:
> The speeds will be afair (value / 65536)_mph. (I know this is definitely correct for openloco pretty sure openrct2 is the same)

This doesn't seem right - in my tests, I found that 65536 seems to be closer to 2mph than 1mph. If I set the car velocity to 655360, the velocity in game is about 21mph. Are you sure the conversion isn't `value/32768`?

I can have the velocity literals in units of mph, but IMO it only really makes sense to do that if those values match up with the displayed velocity in game, and with that conversion factor they differ by a factor of 2.

duncanspumpkin replied:
As said i haven't actually checked for openrct2 i only know openlocos off by heart.

duncanspumpkin replied:
The actual conversion is `value / 147,456`

X123M3-256 replied:
> The actual conversion is `value / 147,456`

So I found [this comment](https://github.com/OpenRCT2/OpenRCT2/pull/16863#discussion_r837382056), where you said that the conversion is 65536/2.25=29127, which agrees with my experiments. I'm not sure where the 147456 came from but I think I'm going to change it to use 29127 as the conversion factor, as that actually results in sensible speeds.

duncanspumpkin replied:
Sorry yeah multiplied it instead of divide was doing things quickly

X123M3-256 commented:
@duncanspumpkin OK, I've moved the code into a separate file and removed the rider_speed_preference field, replacing it with the rider sprite ID. I've also changed the rider speed distribution (normal distribution doesn't really make sense for this, it technically means there's a nonzero chance of a negative speed).

I've left the speed values as integers because I still don't really know how to convert them into mph. But, those values are randomly generated by a script and not really meant to be edited by hand, so is this really a problem?

X123M3-256 commented:
I'm not sure why the build is failing, it worked when I tested it.

X123M3-256 commented on src/openrct2/ride/VehicleRiderControl.cpp:
```diff
@@ -0,0 +1,252 @@
+#include "Track.h"
+#include "TrackData.h"
+#include "TrainManager.h"
+#include "Vehicle.h"
```
It's because I used Vehicle.cpp as a template, I've deleted all the includes that aren't required.

tupaschoal reviewed (commented):
@X123M3-256 can you please mark the conversations as resolved if you've implemented the suggestion? Then we know that the ones open are the ones you still need feedback on

duncanspumpkin reviewed (approved):
@tupaschoal will this be annoying with your mph branch or not? Should just be a find replace but note the factor used.

tupaschoal commented:
Will give a bit of trouble, I'd rather have we merge that one before, if possible

X123M3-256 commented:
What still needs to be done on this?

tupaschoal commented on src/openrct2/ride/VehicleRiderControl.cpp:
```diff
@@ -0,0 +1,226 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2020 OpenRCT2 developers
```
```suggestion
 * Copyright (c) 2014-2022 OpenRCT2 developers
```

tupaschoal commented on src/openrct2/ride/VehicleRiderControl.cpp:
```diff
@@ -0,0 +1,226 @@
+
+    RiderControlSettings riderSettings = riderTable[peep[0].ToUnderlying() & 0xFF];
+
+    // Acceleration to apply when rider is braking
+    int32_t minBrake = (1 << 16);
+
+    // Acceleration to apply when rider is braking hard (used when going much faster than the rider's preferred speed or very
+    // close to the rider in front)
+    int32_t maxBrake = (12 << 16);
+
+    // Velocity above which riders will attempt to maintain separation from the vehicle in front
+    int32_t minFollowVelocity = MPH(4);
+
+    // Minimum separation distance that riders will allow (regardless of followDistance)
+    int32_t minFollowDistance = 32;
```
Since these are not updated on the function and are based on literals, they should probably be `constexpr` variables defined at the top of the file.

tupaschoal reviewed (commented):
> What still needs to be done on this?

I was waiting for #16863 to be merged, but it likely will be more problematic than anticipated. Nonetheless, I still found a few improvements that can be made.
